File: src/Sems.Api/Shared/Events/DomainEvents.cs
namespace Sems.Api.Shared.Events;

/// <summary>
/// Contrato comun de todo evento de dominio: sabe cuando ocurrio y a que
/// usuario pertenece.
/// </summary>
public interface IDomainEvent
{
    Guid UserId { get; }

    DateTime OccurredAt { get; }
}

/// <summary>
/// Eventos que cruzan bounded contexts.
///
/// <para>Sustituyen a los topics de Kafka del disenio anterior. Cada topic paso
/// a ser un tipo de evento; la entrega ocurre dentro del mismo proceso, despues
/// de que la transaccion del emisor confirme.</para>
///
/// <para>Equivalencia con el disenio de microservicios:</para>
/// <code>
///   iam.events            -> UserRegistered, UserLoggedIn, VerificationRequested, PasswordResetRequested
///   device.events         -> DeviceRegistered, DeviceLinked, DeviceUnlinked, DeviceStatusUpdated
///   energy.events         -> ReadingProcessed
///   alerts.events         -> AlertTriggered
///   payments.events       -> PaymentProcessed
///   subscriptions.events  -> SubscriptionChanged
/// </code>
///
/// <para>Los consumidores no conocen al emisor: escuchan el tipo. Si en el
/// futuro se vuelve a un broker, basta anadir un adaptador que reenvie estos
/// mismos eventos, sin tocar emisores ni consumidores.</para>
/// </summary>
public static class DomainEvents
{
    // ------------------------------------------------------------------ iam

    public sealed record UserRegistered(Guid UserId, string EmailAddress, string Role)
        : IDomainEvent
    {
        public DateTime OccurredAt { get; init; } = DateTime.UtcNow;
    }

    public sealed record UserLoggedIn(Guid UserId, string EmailAddress) : IDomainEvent
    {
        public DateTime OccurredAt { get; init; } = DateTime.UtcNow;
    }

    public sealed record RoleAssigned(Guid UserId, string Role) : IDomainEvent
    {
        public DateTime OccurredAt { get; init; } = DateTime.UtcNow;
    }

    /// <summary>Pide al modulo de notificaciones el codigo de verificacion.</summary>
    public sealed record VerificationRequested(Guid UserId, string EmailAddress, string Token)
        : IDomainEvent
    {
        public DateTime OccurredAt { get; init; } = DateTime.UtcNow;
    }

    /// <summary>Pide el enlace de recuperacion de contrasena.</summary>
    public sealed record PasswordResetRequested(Guid UserId, string EmailAddress, string Token)
        : IDomainEvent
    {
        public DateTime OccurredAt { get; init; } = DateTime.UtcNow;
    }

    // -------------------------------------------------------------- devices

    public sealed record DeviceRegistered(Guid UserId, Guid DeviceId, string Name, string Type)
        : IDomainEvent
    {
        public DateTime OccurredAt { get; init; } = DateTime.UtcNow;
    }

    public sealed record DeviceLinked(Guid UserId, Guid DeviceId, Guid BindingId) : IDomainEvent
    {
        public DateTime OccurredAt { get; init; } = DateTime.UtcNow;
    }

    public sealed record DeviceUnlinked(Guid UserId, Guid DeviceId, Guid BindingId) : IDomainEvent
    {
        public DateTime OccurredAt { get; init; } = DateTime.UtcNow;
    }

    public sealed record DeviceStatusUpdated(Guid UserId, Guid DeviceId, string Status)
        : IDomainEvent
    {
        public DateTime OccurredAt { get; init; } = DateTime.UtcNow;
    }

    // --------------------------------------------------------------- energy

    /// <summary>
    /// Lectura del medidor ya validada y almacenada. La escuchan analytics
    /// (rankings y proyecciones) y alerts (evaluacion de umbrales).
    /// </summary>
    public sealed record ReadingProcessed(Guid UserId, Guid? DeviceId, Guid? MeterId,
        decimal ConsumptionKwh, DateTime RecordedAt) : IDomainEvent
    {
        public DateTime OccurredAt { get; init; } = DateTime.UtcNow;
    }

    // --------------------------------------------------------------- alerts

    public sealed record AlertTriggered(Guid UserId, Guid AlertId, string AlertType,
        string Severity, string Message) : IDomainEvent
    {
        public DateTime OccurredAt { get; init; } = DateTime.UtcNow;
    }

    // ------------------------------------------------------------- payments

    public sealed record PaymentProcessed(Guid UserId, Guid PaymentId, decimal Amount,
        string Currency, string Status) : IDomainEvent
    {
        public DateTime OccurredAt { get; init; } = DateTime.UtcNow;
    }

    // -------------------------------------------------------- subscriptions

    public sealed record SubscriptionChanged(Guid UserId, Guid SubscriptionId, string PlanName,
        string Status) : IDomainEvent
    {
        public DateTime OccurredAt { get; init; } = DateTime.UtcNow;
    }
}

File: src/main/java/com/sems/shared/events/DomainEvents.java
package com.sems.shared.events;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;

/**
 * Eventos de dominio que cruzan bounded contexts.
 *
 * <p>Sustituyen a los topics de Kafka del disenio anterior. Cada topic paso a ser
 * un tipo de evento; la entrega la hace {@link org.springframework.context.ApplicationEventPublisher}
 * dentro del mismo proceso, despues de que la transaccion del emisor confirme.
 *
 * <p>Equivalencia con el disenio de microservicios:
 * <pre>
 *   iam.events            -> UserRegistered, UserLoggedIn, VerificationRequested, PasswordResetRequested
 *   device.events         -> DeviceRegistered, DeviceLinked, DeviceUnlinked, DeviceStatusUpdated
 *   energy.events         -> ReadingProcessed
 *   alerts.events         -> AlertTriggered
 *   payments.events       -> PaymentProcessed
 *   subscriptions.events  -> SubscriptionChanged
 * </pre>
 *
 * <p>Los consumidores no conocen al emisor: escuchan el tipo. Si en el futuro se
 * vuelve a un broker, basta con anadir un adaptador que reenvie estos mismos
 * eventos, sin tocar ni emisores ni consumidores.
 */
public final class DomainEvents {

    private DomainEvents() {
    }

    /** Contrato comun: todo evento sabe cuando ocurrio y a que usuario pertenece. */
    public interface DomainEvent {
        UUID userId();

        Instant occurredAt();
    }

    // ---------------------------------------------------------------- iam

    public record UserRegistered(UUID userId, String emailAddress, String role, Instant occurredAt)
            implements DomainEvent {
        public UserRegistered(UUID userId, String emailAddress, String role) {
            this(userId, emailAddress, role, Instant.now());
        }
    }

    public record UserLoggedIn(UUID userId, String emailAddress, Instant occurredAt)
            implements DomainEvent {
        public UserLoggedIn(UUID userId, String emailAddress) {
            this(userId, emailAddress, Instant.now());
        }
    }

    public record RoleAssigned(UUID userId, String role, Instant occurredAt)
            implements DomainEvent {
        public RoleAssigned(UUID userId, String role) {
            this(userId, role, Instant.now());
        }
    }

    /** Pide al modulo de notificaciones que envie el codigo de verificacion de cuenta. */
    public record VerificationRequested(UUID userId, String emailAddress, String token, Instant occurredAt)
            implements DomainEvent {
        public VerificationRequested(UUID userId, String emailAddress, String token) {
            this(userId, emailAddress, token, Instant.now());
        }
    }

    /** Pide al modulo de notificaciones que envie el enlace de recuperacion de contrasena. */
    public record PasswordResetRequested(UUID userId, String emailAddress, String token, Instant occurredAt)
            implements DomainEvent {
        public PasswordResetRequested(UUID userId, String emailAddress, String token) {
            this(userId, emailAddress, token, Instant.now());
        }
    }

    // ------------------------------------------------------------ devices

    public record DeviceRegistered(UUID userId, UUID deviceId, String name, String type, Instant occurredAt)
            implements DomainEvent {
        public DeviceRegistered(UUID userId, UUID deviceId, String name, String type) {
            this(userId, deviceId, name, type, Instant.now());
        }
    }

    public record DeviceLinked(UUID userId, UUID deviceId, UUID bindingId, Instant occurredAt)
            implements DomainEvent {
        public DeviceLinked(UUID userId, UUID deviceId, UUID bindingId) {
            this(userId, deviceId, bindingId, Instant.now());
        }
    }

    public record DeviceUnlinked(UUID userId, UUID deviceId, UUID bindingId, Instant occurredAt)
            implements DomainEvent {
        public DeviceUnlinked(UUID userId, UUID deviceId, UUID bindingId) {
            this(userId, deviceId, bindingId, Instant.now());
        }
    }

    public record DeviceStatusUpdated(UUID userId, UUID deviceId, String status, Instant occurredAt)
            implements DomainEvent {
        public DeviceStatusUpdated(UUID userId, UUID deviceId, String status) {
            this(userId, deviceId, status, Instant.now());
        }
    }

    // ------------------------------------------------------------- energy

    /**
     * Una lectura del medidor ya validada y almacenada.
     * La escuchan analytics (para rankings y proyecciones) y alerts (para umbrales).
     */
    public record ReadingProcessed(UUID userId, UUID deviceId, UUID meterId,
                                   BigDecimal consumptionKwh, Instant recordedAt, Instant occurredAt)
            implements DomainEvent {
        public ReadingProcessed(UUID userId, UUID deviceId, UUID meterId,
                                BigDecimal consumptionKwh, Instant recordedAt) {
            this(userId, deviceId, meterId, consumptionKwh, recordedAt, Instant.now());
        }
    }

    // ------------------------------------------------------------- alerts

    public record AlertTriggered(UUID userId, UUID alertId, String alertType,
                                 String severity, String message, Instant occurredAt)
            implements DomainEvent {
        public AlertTriggered(UUID userId, UUID alertId, String alertType, String severity, String message) {
            this(userId, alertId, alertType, severity, message, Instant.now());
        }
    }

    // ----------------------------------------------------------- payments

    public record PaymentProcessed(UUID userId, UUID paymentId, BigDecimal amount,
                                   String currency, String status, Instant occurredAt)
            implements DomainEvent {
        public PaymentProcessed(UUID userId, UUID paymentId, BigDecimal amount, String currency, String status) {
            this(userId, paymentId, amount, currency, status, Instant.now());
        }
    }

    // ------------------------------------------------------ subscriptions

    public record SubscriptionChanged(UUID userId, UUID subscriptionId, String planName,
                                      String status, Instant occurredAt)
            implements DomainEvent {
        public SubscriptionChanged(UUID userId, UUID subscriptionId, String planName, String status) {
            this(userId, subscriptionId, planName, status, Instant.now());
        }
    }
}
